File: root.go
package tempilecore

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type Root struct {
	Childs []Node
}

func (r *Root) ResolveIncludes(srcPath string) error {
	err := resolveIncludes(&r.Childs, srcPath)
	if err != nil {
		return err
	}

	return nil
}

func (r *Root) MatchSlotsAndContents() {
	includeNodes := collectIncludes(&r.Childs)
	slotsMap := make(map[string]map[string]*SlotNode)
	contentsMap := make(map[string]map[string]*ContentNode)

	for _, imp := range includeNodes {
		slots, contents := collectSlotsAndContents(&imp.Childs)
		slotsMap[imp.CtxId] = slots
		contentsMap[imp.CtxId] = contents
	}

	for k, v := range contentsMap {
		for innerK, innerV := range v {
			slot, ok := slotsMap[k][innerK]
			if ok {
				slot.Childs = append(slot.Childs, innerV.Childs...)
			}
		}
	}

	unwrapIncludes(&r.Childs)
	unwrapSlots(&r.Childs)
}

func unwrapIncludes(nodes *[]Node) {
	for i := 0; i < len(*nodes); i++ {
		n := (*nodes)[i]

		if n.Type() == NodeInclude {
			includeNode := n.(*IncludeNode)
			unwrapIncludes(&includeNode.Childs)
			*nodes = append(append((*nodes)[:i], includeNode.Childs...), (*nodes)[i+1:]...)
		} else {
			childs := n.GetChilds()
			if childs != nil {
				unwrapIncludes(childs)
			}
		}
	}
}

func unwrapSlots(nodes *[]Node) {
	for i := 0; i < len(*nodes); i++ {
		n := (*nodes)[i]

		if n.Type() == NodeSlot {
			slotNode := n.(*SlotNode)
			unwrapSlots(&slotNode.Childs)
			*nodes = append(append((*nodes)[:i], slotNode.Childs...), (*nodes)[i+1:]...)
		} else {
			childs := n.GetChilds()
			if childs != nil {
				unwrapSlots(childs)
			}
		}
	}
}

func collectIncludes(nodes *[]Node) []*IncludeNode {
	includeNodes := []*IncludeNode{}

	for _, n := range *nodes {
		if n.Type() == NodeInclude {
			includeNode := n.(*IncludeNode)
			includeNodes = append(includeNodes, includeNode)
		}
		if childs := n.GetChilds(); childs != nil {
			includeNodes = append(includeNodes, collectIncludes(childs)...)
		}
	}

	return includeNodes
}

func collectSlotsAndContents(nodes *[]Node) (map[string]*SlotNode, map[string]*ContentNode) {
	slots := make(map[string]*SlotNode)
	contents := make(map[string]*ContentNode)
	newNodes := (*nodes)[:0]

	for _, n := range *nodes {
		if n.Type() == NodeSlot {
			newNodes = append(newNodes, n)
			slotNode := n.(*SlotNode)
			slots[slotNode.Name] = slotNode
		} else if n.Type() == NodeContent {
			contentNode := n.(*ContentNode)
			contents[contentNode.Name] = contentNode
		} else {
			newNodes = append(newNodes, n)
		}

		childs := n.GetChilds()
		if childs != nil {
			s, c := collectSlotsAndContents(childs)
			for sk, sv := range s {
				slots[sk] = sv
			}

			for ck, cv := range c {
				contents[ck] = cv
			}
		}
	}

	*nodes = newNodes
	return slots, contents
}

func resolveIncludes(nodes *[]Node, srcPath string) error {
	for _, n := range *nodes {
		if n.Type() == NodeInclude {
			includeNode := n.(*IncludeNode)

			pieces := strings.Split(filepath.Join(srcPath, includeNode.Path), "/")
			fileName := pieces[len(pieces)-1]

			fileByte, err := os.ReadFile(filepath.Join(srcPath, includeNode.Path))
			if err != nil {
				errMessage := fmt.Sprintf("%v \n file: %s line: %d col: %d", err, includeNode.Pos.FileName, includeNode.Pos.Line, includeNode.Pos.Column)
				return errors.New(errMessage)
			}

			ast, err := Parse(string(fileByte), fileName)
			if err != nil {
				return err
			}

			resolveIncludes(&ast.Childs, srcPath)

			for i := 0; i < len(includeNode.Childs); i++ {
				n := includeNode.Childs[i]

				if n.Type() != NodeContent {
					includeNode.Childs = append(includeNode.Childs[:i], includeNode.Childs[i+1:]...)
					i = i - 1
				}
			}

			resolveIncludes(&includeNode.Childs, srcPath)

			includeNode.Childs = append(ast.Childs, includeNode.Childs...)

		} else {
			childs := n.GetChilds()
			if childs != nil {
				resolveIncludes(childs, srcPath)
			}
		}
	}

	return nil
}

File: src/ast/base/Root.ts
import path from "node:path";
import NodeType from "./node-types.js";
import Node from "./Node.js";
import fs from "fs";
import { ContentNode, IncludeNode, SlotNode } from "../nodes.js";
import { parse } from "../../parser/parser.js";

type slotsMap = Map<string, Map<string, SlotNode>>;
type contentsMap = Map<string, Map<string, ContentNode>>;

class Root {
    public children: Node[];
    public fileName: string;

    constructor(children: Node[], fileName: string) {
        this.children = children;
        this.fileName = fileName;
    }

    public resolveIncludes(srcPath: string) {
        this.resolveIncludesRecursive(this.children, srcPath);
    }

    private resolveIncludesRecursive(nodes: Node[], srcPath: string) {
        for (const child of nodes) {
            if (child.type === NodeType.Include) {
                const includeNode: IncludeNode = child as IncludeNode;
                const pos = includeNode.path.pos;

                if (path.basename(includeNode.path.value) === this.fileName) {
                    throw new Error(
                        `Circular include detected: file "${this.fileName}" tries to include itself at ${includeNode.path.value}
                        File: ${pos?.fileName} Line: ${pos?.startLine} Col: ${pos?.startCol}  
                        `
                    );
                }
                const filePath: string = path.join(srcPath, includeNode.path.value);
                let file: string = "";
                try {
                    file = fs.readFileSync(filePath, "utf-8");
                } catch (err) {
                    throw new Error(
                        `Failed to read file at path "${filePath}". Error: ${(err as Error).message} 
                        File: ${pos?.fileName} Line: ${pos?.startLine} Col: ${pos?.startCol}
                        `
                    );
                }

                const parsedRoot = parse(file, filePath);

                parsedRoot.resolveIncludes(srcPath);
                this.resolveIncludesRecursive(includeNode.children, srcPath);

                includeNode.children = [...parsedRoot.children, ...includeNode.children];
            } else {
                this.resolveIncludesRecursive(child.getChildren(), srcPath);
            }
        }
    }

    public matchSlotsAndContents() {
        const includeNodes = this.collectIncludes(this.children);
        const slots: slotsMap = new Map();
        const contents: contentsMap = new Map();

        for (const iNode of includeNodes) {
            slots.set(iNode.ctxId, new Map());
            contents.set(iNode.ctxId, new Map());

            this.collectSlotsAndContents(iNode.children, iNode.ctxId, slots, contents);
        }

        for (const [k, v] of contents) {
            for (const [innerK, innerV] of v) {
                const slot = slots.get(k)?.get(innerK);
                slot?.children.push(...innerV.children);
            }
        }

        this.unwrapSlotsAndIncludes(this.children);
    }

    private collectIncludes(nodes: Node[]): IncludeNode[] {
        const includeNodes: IncludeNode[] = [];

        for (const node of nodes) {
            if (node.type === NodeType.Include) {
                includeNodes.push(node as IncludeNode);
            }

            includeNodes.push(...this.collectIncludes(node.getChildren()));
        }

        return includeNodes;
    }

    private collectSlotsAndContents(nodes: Node[], ctxId: string, slots: slotsMap, contents: contentsMap) {
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];

            if (node.type === NodeType.Slot) {
                slots.get(ctxId)?.set((node as SlotNode).name, node as SlotNode);
            } else if (node.type === NodeType.Content) {
                contents.get(ctxId)?.set((node as ContentNode).name, node as ContentNode);
                this.collectSlotsAndContents(node.getChildren(), ctxId, slots, contents);
                nodes.splice(i, 1);
                i--;
                continue;
            }
            this.collectSlotsAndContents(node.getChildren(), ctxId, slots, contents);
        }
    }

    private unwrapSlotsAndIncludes(nodes: Node[]) {
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];

            const isUnwrap =
                node.type === NodeType.Include ||
                node.type === NodeType.Slot;

            const children = node.getChildren();

            if (isUnwrap) {
                this.unwrapSlotsAndIncludes(children);

                nodes.splice(i, 1, ...children);

                i += children.length - 1;
            } else {
                this.unwrapSlotsAndIncludes(children);
            }
        }
    }

}

export default Root;